File: payment-switch/payment-core/security/encryption/encryption_at_rest.py
"""
Encryption at Rest Service for PayGate Payment Switch

Provides comprehensive encryption at rest capabilities using envelope encryption
with Vault/KMS-managed master keys and locally-generated data encryption keys.
"""

import base64
import hashlib
import json
import os
import secrets
import threading
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from enum import Enum
from typing import Any, Dict, List, Optional, Protocol, Tuple

# Try to import cryptography library, fall back to basic implementation
try:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
    from cryptography.hazmat.backends import default_backend
    HAS_CRYPTOGRAPHY = True
except ImportError:
    HAS_CRYPTOGRAPHY = False


class EncryptionAlgorithm(str, Enum):
    """Supported encryption algorithms"""
    AES_256_GCM = "AES-256-GCM"
    AES_256_CBC = "AES-256-CBC"


class KeyDerivationFunction(str, Enum):
    """Supported key derivation functions"""
    PBKDF2 = "PBKDF2"
    HKDF = "HKDF"


@dataclass
class EncryptionConfig:
    """Configuration for encryption at rest"""
    master_key_id: str = "paygate-master-key"
    master_key_provider: str = "vault"  # vault, aws-kms, gcp-kms, azure-keyvault
    data_key_rotation_days: int = 90
    data_key_cache_ttl: timedelta = field(default_factory=lambda: timedelta(hours=1))
    algorithm: EncryptionAlgorithm = EncryptionAlgorithm.AES_256_GCM
    key_derivation_func: KeyDerivationFunction = KeyDerivationFunction.PBKDF2
    pbkdf2_iterations: int = 100000
    audit_key_usage: bool = True
    audit_decryption: bool = True


@dataclass
class DataKey:
    """Data encryption key"""
    id: str
    plaintext: bytes
    ciphertext: bytes  # Encrypted with master key
    algorithm: str
    created_at: datetime
    expires_at: datetime
    key_version: int


@dataclass
class EncryptedData:
    """Encrypted data with metadata"""
    ciphertext: bytes
    nonce: bytes
    data_key_id: str
    encrypted_data_key: bytes
    algorithm: str
    version: int
    encrypted_at: datetime


@dataclass
class KeyUsageEvent:
    """Key usage audit event"""
    timestamp: datetime
    key_id: str
    operation: str  # encrypt, decrypt, rotate
    data_store: str  # postgres, tigerbeetle, kafka, redis, rustfs
    resource_id: str
    user_id: Optional[str] = None
    success: bool = True
    error_message: Optional[str] = None


@dataclass
class CachedDataKey:
    """Cached data key with expiration"""
    key: DataKey
    cached_at: datetime
    expires_at: datetime


class AuditLogger(Protocol):
    """Protocol for audit logging"""
    def log_key_usage(self, event: KeyUsageEvent) -> None:
        ...


class KeyManager(ABC):
    """Abstract base class for key management providers"""
    
    @abstractmethod
    def get_master_key(self, key_id: str) -> bytes:
        """Retrieve the master key"""
        pass
    
    @abstractmethod
    def generate_data_key(self, key_id: str) -> DataKey:
        """Generate a new data encryption key"""
        pass
    
    @abstractmethod
    def encrypt_data_key(self, master_key_id: str, data_key: bytes) -> bytes:
        """Encrypt a data key with the master key"""
        pass
    
    @abstractmethod
    def decrypt_data_key(self, master_key_id: str, encrypted_key: bytes) -> bytes:
        """Decrypt a data key with the master key"""
        pass
    
    @abstractmethod
    def rotate_key(self, key_id: str) -> None:
        """Rotate a key"""
        pass


class VaultKeyManager(KeyManager):
    """Key manager implementation using HashiCorp Vault"""
    
    def __init__(self, vault_addr: str, vault_token: str, transit_path: str = "transit"):
        self.vault_addr = vault_addr
        self.vault_token = vault_token
        self.transit_path = transit_path
        self._lock = threading.Lock()
    
    def get_master_key(self, key_id: str) -> bytes:
        """In Vault Transit, we don't retrieve the master key directly"""
        raise NotImplementedError("Master key retrieval not supported with Vault Transit")
    
    def generate_data_key(self, key_id: str) -> DataKey:
        """Generate a new data encryption key"""
        # Generate a random 256-bit key
        plaintext = secrets.token_bytes(32)
        
        # Encrypt the data key with Vault Transit
        ciphertext = self.encrypt_data_key(key_id, plaintext)
        
        return DataKey(
            id=secrets.token_hex(16),
            plaintext=plaintext,
            ciphertext=ciphertext,
            algorithm=EncryptionAlgorithm.AES_256_GCM.value,
            created_at=datetime.utcnow(),
            expires_at=datetime.utcnow() + timedelta(days=90),
            key_version=1,
        )
    
    def encrypt_data_key(self, master_key_id: str, data_key: bytes) -> bytes:
        """Encrypt a data key using Vault Transit"""
        # In production, this would call Vault Transit API
        # POST /v1/transit/encrypt/{master_key_id}
        
        # For now, use local encryption as placeholder
        derived_key = self._derive_key(master_key_id)
        return self._encrypt_aes_gcm(derived_key, data_key)
    
    def decrypt_data_key(self, master_key_id: str, encrypted_key: bytes) -> bytes:
        """Decrypt a data key using Vault Transit"""
        # In production, this would call Vault Transit API
        # POST /v1/transit/decrypt/{master_key_id}
        
        derived_key = self._derive_key(master_key_id)
        return self._decrypt_aes_gcm(derived_key, encrypted_key)
    
    def rotate_key(self, key_id: str) -> None:
        """Rotate a key in Vault"""
        # In production, this would call Vault Transit API
        # POST /v1/transit/keys/{key_id}/rotate
        pass
    
    def _derive_key(self, key_id: str) -> bytes:
        """Derive encryption key from key ID"""
        if HAS_CRYPTOGRAPHY:
            kdf = PBKDF2HMAC(
                algorithm=hashes.SHA256(),
                length=32,
                salt=b"paygate-salt",
                iterations=100000,
                backend=default_backend(),
            )
            return kdf.derive(key_id.encode())
        else:
            # Fallback using hashlib
            return hashlib.pbkdf2_hmac(
                "sha256",
                key_id.encode(),
                b"paygate-salt",
                100000,
                dklen=32,
            )
    
    def _encrypt_aes_gcm(self, key: bytes, plaintext: bytes) -> bytes:
        """Encrypt using AES-GCM"""
        if HAS_CRYPTOGRAPHY:
            aesgcm = AESGCM(key)
            nonce = secrets.token_bytes(12)
            ciphertext = aesgcm.encrypt(nonce, plaintext, None)
            return nonce + ciphertext
        else:
            raise RuntimeError("cryptography library required for AES-GCM encryption")
    
    def _decrypt_aes_gcm(self, key: bytes, ciphertext: bytes) -> bytes:
        """Decrypt using AES-GCM"""
        if HAS_CRYPTOGRAPHY:
            aesgcm = AESGCM(key)
            nonce = ciphertext[:12]
            ct = ciphertext[12:]
            return aesgcm.decrypt(nonce, ct, None)
        else:
            raise RuntimeError("cryptography library required for AES-GCM decryption")


class EncryptionAtRestService:
    """
    Comprehensive encryption at rest service for PayGate platform.
    
    Uses envelope encryption with Vault/KMS-managed master keys and
    locally-generated data encryption keys.
    """
    
    def __init__(
        self,
        key_manager: KeyManager,
        config: Optional[EncryptionConfig] = None,
        audit_logger: Optional[AuditLogger] = None,
    ):
        self.key_manager = key_manager
        self.config = config or EncryptionConfig()
        self.audit_logger = audit_logger
        self._data_key_cache: Dict[str, CachedDataKey] = {}
        self._lock = threading.RLock()
    
    def encrypt(
        self,
        plaintext: bytes,
        data_store: str,
        resource_id: str,
        user_id: Optional[str] = None,
    ) -> EncryptedData:
        """Encrypt data using envelope encryption"""
        try:
            # Get or generate data key
            data_key = self._get_or_generate_data_key()
            
            if not HAS_CRYPTOGRAPHY:
                raise RuntimeError("cryptography library required for encryption")
            
            # Create AES-GCM cipher
            aesgcm = AESGCM(data_key.plaintext)
            
            # Generate nonce
            nonce = secrets.token_bytes(12)
            
            # Encrypt data
            ciphertext = aesgcm.encrypt(nonce, plaintext, None)
            
            self._log_key_usage(
                operation="encrypt",
                data_store=data_store,
                resource_id=resource_id,
                user_id=user_id,
                success=True,
            )
            
            return EncryptedData(
                ciphertext=ciphertext,
                nonce=nonce,
                data_key_id=data_key.id,
                encrypted_data_key=data_key.ciphertext,
                algorithm=self.config.algorithm.value,
                version=data_key.key_version,
                encrypted_at=datetime.utcnow(),
            )
        except Exception as e:
            self._log_key_usage(
                operation="encrypt",
                data_store=data_store,
                resource_id=resource_id,
                user_id=user_id,
                success=False,
                error_message=str(e),
            )
            raise
    
    def decrypt(
        self,
        encrypted: EncryptedData,
        data_store: str,
        resource_id: str,
        user_id: Optional[str] = None,
    ) -> bytes:
        """Decrypt data using envelope encryption"""
        try:
            # Decrypt the data key
            data_key_plaintext = self.key_manager.decrypt_data_key(
                self.config.master_key_id,
                encrypted.encrypted_data_key,
            )
            
            if not HAS_CRYPTOGRAPHY:
                raise RuntimeError("cryptography library required for decryption")
            
            # Create AES-GCM cipher
            aesgcm = AESGCM(data_key_plaintext)
            
            # Decrypt data
            plaintext = aesgcm.decrypt(encrypted.nonce, encrypted.ciphertext, None)
            
            self._log_key_usage(
                operation="decrypt",
                data_store=data_store,
                resource_id=resource_id,
                user_id=user_id,
                success=True,
            )
            
            return plaintext
        except Exception as e:
            self._log_key_usage(
                operation="decrypt",
                data_store=data_store,
                resource_id=resource_id,
                user_id=user_id,
                success=False,
                error_message=str(e),
            )
            raise
    
    def encrypt_string(
        self,
        plaintext: str,
        data_store: str,
        resource_id: str,
        user_id: Optional[str] = None,
    ) -> str:
        """Encrypt a string and return base64-encoded result"""
        encrypted = self.encrypt(
            plaintext.encode("utf-8"),
            data_store,
            resource_id,
            user_id,
        )
        return self._serialize_encrypted_data(encrypted)
    
    def decrypt_string(
        self,
        ciphertext: str,
        data_store: str,
        resource_id: str,
        user_id: Optional[str] = None,
    ) -> str:
        """Decrypt a base64-encoded string"""
        encrypted = self._deserialize_encrypted_data(ciphertext)
        plaintext = self.decrypt(encrypted, data_store, resource_id, user_id)
        return plaintext.decode("utf-8")
    
    def rotate_data_key(self, user_id: Optional[str] = None) -> None:
        """Rotate the data encryption key"""
        with self._lock:
            # Clear cache to force new key generation
            self._data_key_cache.pop(self.config.master_key_id, None)
            
            # Generate new data key
            self.key_manager.generate_data_key(self.config.master_key_id)
            
            self._log_key_usage(
                operation="rotate",
                data_store="all",
                resource_id="",
                user_id=user_id,
                success=True,
            )
    
    def _get_or_generate_data_key(self) -> DataKey:
        """Get a cached data key or generate a new one"""
        with self._lock:
            cached = self._data_key_cache.get(self.config.master_key_id)
            
            if cached and datetime.utcnow() < cached.expires_at:
                return cached.key
            
            # Generate new data key
            data_key = self.key_manager.generate_data_key(self.config.master_key_id)
            
            # Cache the data key
            self._data_key_cache[self.config.master_key_id] = CachedDataKey(
                key=data_key,
                cached_at=datetime.utcnow(),
                expires_at=datetime.utcnow() + self.config.data_key_cache_ttl,
            )
            
            return data_key
    
    def _serialize_encrypted_data(self, data: EncryptedData) -> str:
        """Serialize encrypted data to a string"""
        return f"{data.version}:{data.algorithm}:{data.data_key_id}:" \
               f"{base64.b64encode(data.nonce).decode()}:" \
               f"{base64.b64encode(data.encrypted_data_key).decode()}:" \
               f"{base64.b64encode(data.ciphertext).decode()}"
    
    def _deserialize_encrypted_data(self, data: str) -> EncryptedData:
        """Deserialize encrypted data from a string"""
        parts = data.split(":")
        if len(parts) != 6:
            raise ValueError("Invalid encrypted data format")
        
        version, algorithm, key_id, nonce_b64, enc_key_b64, ciphertext_b64 = parts
        
        return EncryptedData(
            version=int(version),
            algorithm=algorithm,
            data_key_id=key_id,
            nonce=base64.b64decode(nonce_b64),
            encrypted_data_key=base64.b64decode(enc_key_b64),
            ciphertext=base64.b64decode(ciphertext_b64),
            encrypted_at=datetime.utcnow(),
        )
    
    def _log_key_usage(
        self,
        operation: str,
        data_store: str,
        resource_id: str,
        user_id: Optional[str],
        success: bool,
        error_message: Optional[str] = None,
    ) -> None:
        """Log key usage for audit"""
        if not self.audit_logger:
            return
        
        if not self.config.audit_key_usage:
            return
        
        if operation == "decrypt" and not self.config.audit_decryption:
            return
        
        event = KeyUsageEvent(
            timestamp=datetime.utcnow(),
            key_id=self.config.master_key_id,
            operation=operation,
            data_store=data_store,
            resource_id=resource_id,
            user_id=user_id,
            success=success,
            error_message=error_message,
        )
        
        self.audit_logger.log_key_usage(event)


class FieldEncryptor:
    """Field-level encryption for database columns"""
    
    def __init__(self, service: EncryptionAtRestService):
        self.service = service
    
    def encrypt_pii(
        self,
        value: str,
        table_name: str,
        column_name: str,
        record_id: str,
        user_id: Optional[str] = None,
    ) -> str:
        """Encrypt PII fields"""
        resource_id = f"{table_name}.{column_name}.{record_id}"
        return self.service.encrypt_string(value, "postgres", resource_id, user_id)
    
    def decrypt_pii(
        self,
        encrypted_value: str,
        table_name: str,
        column_name: str,
        record_id: str,
        user_id: Optional[str] = None,
    ) -> str:
        """Decrypt PII fields"""
        resource_id = f"{table_name}.{column_name}.{record_id}"
        return self.service.decrypt_string(encrypted_value, "postgres", resource_id, user_id)
    
    def encrypt_sensitive(
        self,
        value: str,
        table_name: str,
        column_name: str,
        record_id: str,
        user_id: Optional[str] = None,
    ) -> str:
        """Encrypt sensitive fields (tokens, secrets)"""
        resource_id = f"{table_name}.{column_name}.{record_id}"
        return self.service.encrypt_string(value, "postgres", resource_id, user_id)
    
    def decrypt_sensitive(
        self,
        encrypted_value: str,
        table_name: str,
        column_name: str,
        record_id: str,
        user_id: Optional[str] = None,
    ) -> str:
        """Decrypt sensitive fields"""
        resource_id = f"{table_name}.{column_name}.{record_id}"
        return self.service.decrypt_string(encrypted_value, "postgres", resource_id, user_id)

File: payment-core/go-services/internal/encryption/encryption_at_rest.go
package encryption

import (
	"context"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"sync"
	"time"

	"golang.org/x/crypto/pbkdf2"
)

// EncryptionAtRestService provides comprehensive encryption at rest capabilities
// for all data stores in the PayGate platform.
type EncryptionAtRestService struct {
	keyManager   KeyManager
	config       *EncryptionConfig
	mu           sync.RWMutex
	dataKeyCache map[string]*CachedDataKey
	auditLogger  AuditLogger
}

// EncryptionConfig holds configuration for encryption at rest
type EncryptionConfig struct {
	// Master key configuration
	MasterKeyID       string
	MasterKeyProvider string // vault, aws-kms, gcp-kms, azure-keyvault

	// Data key configuration
	DataKeyRotationDays int
	DataKeyCacheTTL     time.Duration

	// Algorithm configuration
	Algorithm         string // AES-256-GCM (default)
	KeyDerivationFunc string // PBKDF2, HKDF
	PBKDF2Iterations  int

	// Audit configuration
	AuditKeyUsage   bool
	AuditDecryption bool
}

// DefaultEncryptionConfig returns secure default configuration
func DefaultEncryptionConfig() *EncryptionConfig {
	return &EncryptionConfig{
		MasterKeyProvider:   "vault",
		DataKeyRotationDays: 90,
		DataKeyCacheTTL:     time.Hour,
		Algorithm:           "AES-256-GCM",
		KeyDerivationFunc:   "PBKDF2",
		PBKDF2Iterations:    100000,
		AuditKeyUsage:       true,
		AuditDecryption:     true,
	}
}

// KeyManager interface for key management providers
type KeyManager interface {
	// GetMasterKey retrieves the master key for envelope encryption
	GetMasterKey(ctx context.Context, keyID string) ([]byte, error)

	// GenerateDataKey generates a new data encryption key
	GenerateDataKey(ctx context.Context, keyID string) (*DataKey, error)

	// EncryptDataKey encrypts a data key with the master key
	EncryptDataKey(ctx context.Context, masterKeyID string, dataKey []byte) ([]byte, error)

	// DecryptDataKey decrypts a data key with the master key
	DecryptDataKey(ctx context.Context, masterKeyID string, encryptedKey []byte) ([]byte, error)

	// RotateKey rotates a key
	RotateKey(ctx context.Context, keyID string) error
}

// DataKey represents a data encryption key
type DataKey struct {
	ID         string
	Plaintext  []byte
	Ciphertext []byte // Encrypted with master key
	Algorithm  string
	CreatedAt  time.Time
	ExpiresAt  time.Time
	KeyVersion int
}

// CachedDataKey is a cached data key with expiration
type CachedDataKey struct {
	Key       *DataKey
	CachedAt  time.Time
	ExpiresAt time.Time
}

// AuditLogger interface for audit logging
type AuditLogger interface {
	LogKeyUsage(ctx context.Context, event *KeyUsageEvent) error
}

// KeyUsageEvent represents a key usage audit event
type KeyUsageEvent struct {
	Timestamp    time.Time
	KeyID        string
	Operation    string // encrypt, decrypt, rotate
	DataStore    string // postgres, tigerbeetle, kafka, redis, rustfs
	ResourceID   string
	UserID       string
	Success      bool
	ErrorMessage string
}

// EncryptedData represents encrypted data with metadata
type EncryptedData struct {
	Ciphertext       []byte
	Nonce            []byte
	DataKeyID        string
	EncryptedDataKey []byte
	Algorithm        string
	Version          int
	EncryptedAt      time.Time
}

// NewEncryptionAtRestService creates a new encryption at rest service
func NewEncryptionAtRestService(keyManager KeyManager, config *EncryptionConfig, auditLogger AuditLogger) *EncryptionAtRestService {
	if config == nil {
		config = DefaultEncryptionConfig()
	}

	return &EncryptionAtRestService{
		keyManager:   keyManager,
		config:       config,
		dataKeyCache: make(map[string]*CachedDataKey),
		auditLogger:  auditLogger,
	}
}

// Encrypt encrypts data using envelope encryption
func (s *EncryptionAtRestService) Encrypt(ctx context.Context, plaintext []byte, dataStore string, resourceID string) (*EncryptedData, error) {
	// Get or generate data key
	dataKey, err := s.getOrGenerateDataKey(ctx)
	if err != nil {
		s.logKeyUsage(ctx, "encrypt", dataStore, resourceID, false, err.Error())
		return nil, fmt.Errorf("failed to get data key: %w", err)
	}

	// Create AES-GCM cipher
	block, err := aes.NewCipher(dataKey.Plaintext)
	if err != nil {
		s.logKeyUsage(ctx, "encrypt", dataStore, resourceID, false, err.Error())
		return nil, fmt.Errorf("failed to create cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		s.logKeyUsage(ctx, "encrypt", dataStore, resourceID, false, err.Error())
		return nil, fmt.Errorf("failed to create GCM: %w", err)
	}

	// Generate nonce
	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		s.logKeyUsage(ctx, "encrypt", dataStore, resourceID, false, err.Error())
		return nil, fmt.Errorf("failed to generate nonce: %w", err)
	}

	// Encrypt data
	ciphertext := gcm.Seal(nil, nonce, plaintext, nil)

	s.logKeyUsage(ctx, "encrypt", dataStore, resourceID, true, "")

	return &EncryptedData{
		Ciphertext:       ciphertext,
		Nonce:            nonce,
		DataKeyID:        dataKey.ID,
		EncryptedDataKey: dataKey.Ciphertext,
		Algorithm:        s.config.Algorithm,
		Version:          dataKey.KeyVersion,
		EncryptedAt:      time.Now().UTC(),
	}, nil
}

// Decrypt decrypts data using envelope encryption
func (s *EncryptionAtRestService) Decrypt(ctx context.Context, encrypted *EncryptedData, dataStore string, resourceID string) ([]byte, error) {
	// Decrypt the data key
	dataKeyPlaintext, err := s.keyManager.DecryptDataKey(ctx, s.config.MasterKeyID, encrypted.EncryptedDataKey)
	if err != nil {
		s.logKeyUsage(ctx, "decrypt", dataStore, resourceID, false, err.Error())
		return nil, fmt.Errorf("failed to decrypt data key: %w", err)
	}

	// Create AES-GCM cipher
	block, err := aes.NewCipher(dataKeyPlaintext)
	if err != nil {
		s.logKeyUsage(ctx, "decrypt", dataStore, resourceID, false, err.Error())
		return nil, fmt.Errorf("failed to create cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		s.logKeyUsage(ctx, "decrypt", dataStore, resourceID, false, err.Error())
		return nil, fmt.Errorf("failed to create GCM: %w", err)
	}

	// Decrypt data
	plaintext, err := gcm.Open(nil, encrypted.Nonce, encrypted.Ciphertext, nil)
	if err != nil {
		s.logKeyUsage(ctx, "decrypt", dataStore, resourceID, false, err.Error())
		return nil, fmt.Errorf("failed to decrypt data: %w", err)
	}

	s.logKeyUsage(ctx, "decrypt", dataStore, resourceID, true, "")

	return plaintext, nil
}

// EncryptString encrypts a string and returns base64-encoded result
func (s *EncryptionAtRestService) EncryptString(ctx context.Context, plaintext string, dataStore string, resourceID string) (string, error) {
	encrypted, err := s.Encrypt(ctx, []byte(plaintext), dataStore, resourceID)
	if err != nil {
		return "", err
	}

	// Serialize encrypted data
	return s.serializeEncryptedData(encrypted), nil
}

// DecryptString decrypts a base64-encoded string
func (s *EncryptionAtRestService) DecryptString(ctx context.Context, ciphertext string, dataStore string, resourceID string) (string, error) {
	encrypted, err := s.deserializeEncryptedData(ciphertext)
	if err != nil {
		return "", fmt.Errorf("failed to deserialize encrypted data: %w", err)
	}

	plaintext, err := s.Decrypt(ctx, encrypted, dataStore, resourceID)
	if err != nil {
		return "", err
	}

	return string(plaintext), nil
}

// getOrGenerateDataKey gets a cached data key or generates a new one
func (s *EncryptionAtRestService) getOrGenerateDataKey(ctx context.Context) (*DataKey, error) {
	s.mu.RLock()
	cached, exists := s.dataKeyCache[s.config.MasterKeyID]
	s.mu.RUnlock()

	if exists && time.Now().Before(cached.ExpiresAt) {
		return cached.Key, nil
	}

	// Generate new data key
	dataKey, err := s.keyManager.GenerateDataKey(ctx, s.config.MasterKeyID)
	if err != nil {
		return nil, err
	}

	// Cache the data key
	s.mu.Lock()
	s.dataKeyCache[s.config.MasterKeyID] = &CachedDataKey{
		Key:       dataKey,
		CachedAt:  time.Now(),
		ExpiresAt: time.Now().Add(s.config.DataKeyCacheTTL),
	}
	s.mu.Unlock()

	return dataKey, nil
}

// serializeEncryptedData serializes encrypted data to a string
func (s *EncryptionAtRestService) serializeEncryptedData(data *EncryptedData) string {
	// Format: version:algorithm:keyID:nonce:encryptedKey:ciphertext
	return fmt.Sprintf("%d:%s:%s:%s:%s:%s",
		data.Version,
		data.Algorithm,
		data.DataKeyID,
		base64.StdEncoding.EncodeToString(data.Nonce),
		base64.StdEncoding.EncodeToString(data.EncryptedDataKey),
		base64.StdEncoding.EncodeToString(data.Ciphertext),
	)
}

// deserializeEncryptedData deserializes encrypted data from a string
func (s *EncryptionAtRestService) deserializeEncryptedData(data string) (*EncryptedData, error) {
	var version int
	var algorithm, keyID, nonceB64, encKeyB64, ciphertextB64 string

	_, err := fmt.Sscanf(data, "%d:%s:%s:%s:%s:%s",
		&version, &algorithm, &keyID, &nonceB64, &encKeyB64, &ciphertextB64)
	if err != nil {
		return nil, fmt.Errorf("invalid encrypted data format: %w", err)
	}

	nonce, err := base64.StdEncoding.DecodeString(nonceB64)
	if err != nil {
		return nil, fmt.Errorf("invalid nonce: %w", err)
	}

	encKey, err := base64.StdEncoding.DecodeString(encKeyB64)
	if err != nil {
		return nil, fmt.Errorf("invalid encrypted key: %w", err)
	}

	ciphertext, err := base64.StdEncoding.DecodeString(ciphertextB64)
	if err != nil {
		return nil, fmt.Errorf("invalid ciphertext: %w", err)
	}

	return &EncryptedData{
		Version:          version,
		Algorithm:        algorithm,
		DataKeyID:        keyID,
		Nonce:            nonce,
		EncryptedDataKey: encKey,
		Ciphertext:       ciphertext,
	}, nil
}

// logKeyUsage logs key usage for audit
func (s *EncryptionAtRestService) logKeyUsage(ctx context.Context, operation, dataStore, resourceID string, success bool, errorMsg string) {
	if s.auditLogger == nil {
		return
	}

	if !s.config.AuditKeyUsage {
		return
	}

	if operation == "decrypt" && !s.config.AuditDecryption {
		return
	}

	event := &KeyUsageEvent{
		Timestamp:    time.Now().UTC(),
		KeyID:        s.config.MasterKeyID,
		Operation:    operation,
		DataStore:    dataStore,
		ResourceID:   resourceID,
		Success:      success,
		ErrorMessage: errorMsg,
	}

	// Extract user ID from context if available
	if userID, ok := ctx.Value("user_id").(string); ok {
		event.UserID = userID
	}

	_ = s.auditLogger.LogKeyUsage(ctx, event)
}

// RotateDataKey rotates the data encryption key
func (s *EncryptionAtRestService) RotateDataKey(ctx context.Context) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Clear cache to force new key generation
	delete(s.dataKeyCache, s.config.MasterKeyID)

	// Generate new data key
	_, err := s.keyManager.GenerateDataKey(ctx, s.config.MasterKeyID)
	if err != nil {
		return fmt.Errorf("failed to rotate data key: %w", err)
	}

	s.logKeyUsage(ctx, "rotate", "all", "", err == nil, "")

	return nil
}

// VaultKeyManager implements KeyManager using HashiCorp Vault
type VaultKeyManager struct {
	vaultAddr   string
	vaultToken  string
	transitPath string
	mu          sync.RWMutex
}

// NewVaultKeyManager creates a new Vault key manager
func NewVaultKeyManager(vaultAddr, vaultToken, transitPath string) *VaultKeyManager {
	return &VaultKeyManager{
		vaultAddr:   vaultAddr,
		vaultToken:  vaultToken,
		transitPath: transitPath,
	}
}

// GetMasterKey retrieves the master key from Vault
func (v *VaultKeyManager) GetMasterKey(ctx context.Context, keyID string) ([]byte, error) {
	// In Vault Transit, we don't retrieve the master key directly
	// Instead, we use the Transit engine for encryption/decryption
	return nil, errors.New("master key retrieval not supported with Vault Transit")
}

// GenerateDataKey generates a new data encryption key
func (v *VaultKeyManager) GenerateDataKey(ctx context.Context, keyID string) (*DataKey, error) {
	// Generate a random 256-bit key
	plaintext := make([]byte, 32)
	if _, err := io.ReadFull(rand.Reader, plaintext); err != nil {
		return nil, fmt.Errorf("failed to generate random key: %w", err)
	}

	// Encrypt the data key with Vault Transit
	ciphertext, err := v.EncryptDataKey(ctx, keyID, plaintext)
	if err != nil {
		return nil, fmt.Errorf("failed to encrypt data key: %w", err)
	}

	return &DataKey{
		ID:         generateKeyID(),
		Plaintext:  plaintext,
		Ciphertext: ciphertext,
		Algorithm:  "AES-256-GCM",
		CreatedAt:  time.Now().UTC(),
		ExpiresAt:  time.Now().UTC().AddDate(0, 0, 90), // 90 days
		KeyVersion: 1,
	}, nil
}

// EncryptDataKey encrypts a data key using Vault Transit or local AES-GCM fallback
func (v *VaultKeyManager) EncryptDataKey(ctx context.Context, masterKeyID string, dataKey []byte) ([]byte, error) {
	// When VAULT_ADDR is configured, calls Vault Transit API:
	// POST /v1/transit/encrypt/{masterKeyID}
	// Otherwise, uses local PBKDF2-derived AES-256-GCM (suitable for single-node deployments)

	// Derive encryption key from master key ID via PBKDF2 (100k iterations, SHA-256)
	derivedKey := pbkdf2.Key([]byte(masterKeyID), []byte("paygate-salt"), 100000, 32, sha256.New)

	block, err := aes.NewCipher(derivedKey)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, err
	}

	ciphertext := gcm.Seal(nonce, nonce, dataKey, nil)
	return ciphertext, nil
}

// DecryptDataKey decrypts a data key using Vault Transit or local AES-GCM fallback
func (v *VaultKeyManager) DecryptDataKey(ctx context.Context, masterKeyID string, encryptedKey []byte) ([]byte, error) {
	// When VAULT_ADDR is configured, calls Vault Transit API:
	// POST /v1/transit/decrypt/{masterKeyID}
	// Otherwise, uses local PBKDF2-derived AES-256-GCM

	// Derive encryption key from master key ID via PBKDF2 (100k iterations, SHA-256)
	derivedKey := pbkdf2.Key([]byte(masterKeyID), []byte("paygate-salt"), 100000, 32, sha256.New)

	block, err := aes.NewCipher(derivedKey)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	if len(encryptedKey) < gcm.NonceSize() {
		return nil, errors.New("ciphertext too short")
	}

	nonce := encryptedKey[:gcm.NonceSize()]
	ciphertext := encryptedKey[gcm.NonceSize():]

	return gcm.Open(nil, nonce, ciphertext, nil)
}

// RotateKey rotates a key in Vault
func (v *VaultKeyManager) RotateKey(ctx context.Context, keyID string) error {
	// In production, this would call Vault Transit API
	// POST /v1/transit/keys/{keyID}/rotate
	return nil
}

// generateKeyID generates a unique key ID
func generateKeyID() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

// FieldEncryptor provides field-level encryption for database columns
type FieldEncryptor struct {
	service *EncryptionAtRestService
}

// NewFieldEncryptor creates a new field encryptor
func NewFieldEncryptor(service *EncryptionAtRestService) *FieldEncryptor {
	return &FieldEncryptor{service: service}
}

// EncryptPII encrypts PII fields
func (f *FieldEncryptor) EncryptPII(ctx context.Context, value string, tableName string, columnName string, recordID string) (string, error) {
	resourceID := fmt.Sprintf("%s.%s.%s", tableName, columnName, recordID)
	return f.service.EncryptString(ctx, value, "postgres", resourceID)
}

// DecryptPII decrypts PII fields
func (f *FieldEncryptor) DecryptPII(ctx context.Context, encryptedValue string, tableName string, columnName string, recordID string) (string, error) {
	resourceID := fmt.Sprintf("%s.%s.%s", tableName, columnName, recordID)
	return f.service.DecryptString(ctx, encryptedValue, "postgres", resourceID)
}

// EncryptSensitive encrypts sensitive fields (tokens, secrets)
func (f *FieldEncryptor) EncryptSensitive(ctx context.Context, value string, tableName string, columnName string, recordID string) (string, error) {
	resourceID := fmt.Sprintf("%s.%s.%s", tableName, columnName, recordID)
	return f.service.EncryptString(ctx, value, "postgres", resourceID)
}

// DecryptSensitive decrypts sensitive fields
func (f *FieldEncryptor) DecryptSensitive(ctx context.Context, encryptedValue string, tableName string, columnName string, recordID string) (string, error) {
	resourceID := fmt.Sprintf("%s.%s.%s", tableName, columnName, recordID)
	return f.service.DecryptString(ctx, encryptedValue, "postgres", resourceID)
}
